torch_xla/experimental/stablehlo_saved_model.py

```diff
@@ -0,0 +1,110 @@
+import copy
+import shutil
+import os
+import re
+import numpy as np
+import torch
+from torch import nn
+import torch_xla
+from torch_xla.core import xla_model as xm
+import tensorflow as tf
+import torchvision
+import torch._dynamo as torchdynamo
+
+from tensorflow.compiler.tf2xla.python import xla as tfxla  # type: ignore[import]
+
+from typing import Tuple, Type, Callable
+
+import sys
+
+
+class SHLOModel:
+
+  def __init__(self, model_bytecode, pos_to_orig_pos, pos_to_param,
+               sample_output):
+    self.model_bytecode = model_bytecode
+    self.pos_to_orig_pos = pos_to_orig_pos
+    self.pos_to_param = pos_to_param
+    self._total_number = len(pos_to_orig_pos) + len(pos_to_param)
+    output_tf = tf.convert_to_tensor(sample_output)
+    self.tout = (output_tf.dtype,)
+    self.sout = (tuple(output_tf.shape),)
+
+  def __call__(self, *args):
+
+    call_args = []
+    for i in range(self._total_number):
+      if i in self.pos_to_orig_pos:
+        call_args.append(args[self.pos_to_orig_pos[i]])
+      else:
+        call_args.append(self.pos_to_param[i])
+
+    return tfxla.call_module(
+        tuple(call_args),
+        version=5,
+        Tout=self.tout,  # dtype information
+        Sout=self.sout,  # Shape information
+        function_list=[],
+        platforms=('CPU',),
+        module=self.model_bytecode,
+    )[0]
+
+
+def export_torch_model(model: torch.nn.Module,
+                       sample_inputs: Tuple[torch.Tensor]):
+  """Convert model into a callable backed by StableHLO.
+    
+    Args:
+        model: torch.nn.Module - a pytorch model
+        sample_inputs: Tuple[torch.Tensor] - The input to this model
+
+    This function will return a callable backed by StableHLO such that,
+
+    model(*sample_inputs) ==  export_torch_model(model, sample_inputs)(*sample_inputs)
+    (up to numerics)
+
+    In other words, returned callable have the same calling convention of the input model, and
+    on the sample input, or inputs sufficiently similar* to sample input,  
+    it is will to return same result as the original model.
+
+    * sufficiently similar input because this function will use tracing to extract the model operations
+      so it might specialize on the shapes of the sample input.
+    
+    For now, model has to only take Tensors as input and has to return 1 tensor as output.
+
+    """
+  xm.mark_step()
```

Comment:
nit: add a comment here to explain why a mark_step is needed(something along the line of "Materialize the computation before this call to make sure that StableHLO captured below only reflect the model passed in")

Reply:
done